src/base_object.h

```diff
@@ -103,12 +110,44 @@ class BaseObject : public MemoryRetainer {
   // refer to `doc/guides/node-postmortem-support.md`
   friend int GenDebugSymbols();
   friend class CleanupHookCallback;
+  template <typename T, bool kIsWeak>
+  friend class BaseObjectPtrImpl;
 
   v8::Global<v8::Object> persistent_handle_;
+
+  // Metadata that is associated with this BaseObject if there are BaseObjectPtr
+  // or BaseObjectWeakPtr references to it.
+  // This object is deleted when the BaseObject itself is destroyed, and there
+  // are no weak references to it.
+  struct PointerData {
+    // Number of BaseObjectPtr instances that refer to this object. If this
+    // is non-zero, the BaseObject is always a GC root and will not be destroyed
+    // during cleanup until the count drops to zero again.
+    unsigned int strong_ptr_count = 0;
+    // Number of BaseObjectWeakPtr instances that refer to this object.
+    unsigned int weak_ptr_count = 0;
+    // Indicates whether MakeWeak() has been called.
+    bool wants_weak_jsobj = false;
+    // Indicates whether Detach() has been called. If that is the case, this
+    // object will be destryoed once the strong pointer count drops to zero.
+    bool is_detached = false;
+    // Reference to the original BaseObject. This is used by weak pointers.
+    BaseObject* self = nullptr;
+  };
+
+  inline bool has_pointer_data() const;
+  // This creates a PointerData struct if none was associated with this
+  // BaseObject before.
+  inline PointerData* pointer_data();
+
+  // Functions that adjust the strong pointer count.
+  inline void decrease_refcount();
+  inline void increase_refcount();
+
   Environment* env_;
+  PointerData* pointer_data_ = nullptr;
```

Comment:
Btw, I’ve considered merging the two members into one tagged pointer (`Environment*` by default, `PointerData*` otherwise + an `Environment*` member in `PointerData`) in order to avoid the extra memory overhead for most BaseObjects, but playing around with similar code and looking at the compiler output I got worried about the added CPU cost, and so I let that be for now.